### PR TITLE
Added PreBuildMerge as a Git Trait

### PR DESCRIPTION
**Pull request by arthurvanduynhoven**
For a Multibranch Pipeline job, we need to be able to pre-merge the Jenkinsfile so that the latest changes can be tested properly. This feature was available before the introductions to Git Traits.

### CODE REVIEW

**Comment by stephenc** [comment]
I think this would be better in an extension plugin rather than a core plugin exposed to everyone. The git plugin is already complex enough, WDYT @MarkEWaite

File: src/main/java/jenkins/plugins/git/traits/PreBuildMergeTrait.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 CloudBees, Inc.
```
Comment (stephenc) [inline src/main/java/jenkins/plugins/git/traits/PreBuildMergeTrait.java]: @arthurvanduynhoven does not work for CloudBees last I checked

Comment (arthurvanduynhoven) [inline src/main/java/jenkins/plugins/git/traits/PreBuildMergeTrait.java]: Sorry -- copy paste issue. I remove the mention of CloudBees.

**Comment by arthurvanduynhoven** [comment]
@stephenc If we move this to an extension plugin, how would you want to handle the deprecation of PreBuildMerge class? As I would assume this logic should be migrated to the new extension plugin.

**Comment by arthurvanduynhoven** [comment]
@stephenc Is this what you are looking for?
https://github.com/arthurvanduynhoven/git-prebuildmerge-trait-plugin

**Comment by stephenc** [comment]
I don't think you need to steal the extension, you should be fine with just the trait

**Comment by arthurvanduynhoven** [comment]
@stephenc The reason I "stole" the extension was to include the required methods 
`equals()` `hashCode()` `toString()` and `getDisplayName()`
These changes are not required in the core `git-plugin` as from what I understand the changes are only really needed for the trait and in the end you want to remove `PreBuildMerge` from the core `git-plugin`?

But if you think I should open a new MR for these changes in the core `git-plugin` I can and then I will make the appropriate changes in the trait code.

**Comment by stephenc** [comment]
I think the changes to the extension (which is currently in the git plugin) should live in the git plugin.

A second thing, Your extension plugin should ensure that your extension is limited to GitSCMSource and doesn't show for GitHub or the others as it will conflict with PR merges

**Comment by arthurvanduynhoven** [comment]
(for Reference) I have added MR https://github.com/jenkinsci/git-plugin/pull/517 for the core Git Changes.

For the second item to restrict only to GitSCMSource, doesn't `GitSCMExtensionTrait` make it only for `GitSCMSource`? Or do I need to change that to `SCMSourceTrait` and change `DescriptorImpl` to the following:

```
/**
     * Our {@link hudson.model.Descriptor}.
     */
    @Extension
    @Discovery
    public static class DescriptorImpl extends SCMSourceTraitDescriptor {

        /**
         * {@inheritDoc}
         */
        @Override
        public String getDisplayName() {
            return "Merge before build";
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public Class<? extends SCMBuilder> getBuilderClass() {
            return GitSCMBuilder.class;
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public Class<? extends SCMSourceContext> getContextClass() {
            return GitSCMSourceContext.class;
        }

        /**
         * {@inheritDoc}
         */
        @Override
        public Class<? extends SCMSource> getSourceClass() {
            return GitSCMSource.class;
        }
    }
```

**Comment by stephenc** [comment]
You can just override the methods in your trait's descriptor

**Comment by stephenc** [comment]
IOW in your [`GitSCMExtentionDescriptor`](https://github.com/arthurvanduynhoven/git-prebuildmerge-trait-plugin/blob/master/src/main/java/jenkins/plugins/git/traits/PreBuildMergeTrait.java#L51) impl just add

```
/**
         * {@inheritDoc}
         */
        @Override
        public Class<? extends SCMSource> getSourceClass() {
            return GitSCMSource.class;
        }
```

**Comment by arthurvanduynhoven** [comment]
@stephenc Thanks for your help. I will push the latest changes to the new repo.
Do you know when Git-Plugin (3.5.1?) will be released next? As I will have a dependency on this newly release Git-Plugin due to the PreBuildMerge class changes.

**Comment by arthurvanduynhoven** [comment]
Closing as not needed anymore.

For further tracking please refer to the following items:
Git-Plugin MR #517 
Plugin Hosting Request: https://issues.jenkins-ci.org/browse/HOSTING-380

**Comment by MarkEWaite** [comment]
Git plugin 3.5.1 released 5 Aug 2017.

**Pull request closed without merging**